keps/sig-api-machinery/20190612-crd-group-protection.md

```diff
@@ -0,0 +1,127 @@
+---
+title: k8s.io Group Protection
+authors:
+  - "@deads2k"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-api-machinery
+  - sig-architecture
+reviewers:
+  - "@sttts"
+  - "@jpbetz"
+  - "@liggitt"
+approvers:
+  - "@liggitt"
+  - "@sttts"
+editor:
+  name: "@deads2k"
+creation-date: 2019-06-12
+last-updated: 2019-06-26
+status: implementable
+---
+
+# k8s.io Group Protection
+
+## Table of Contents
+
+## Summary
+
+API groups are organized by namespace, similar to java packages.  `authorization.k8s.io` is one example.  When users create
+CRDs, they get to specify an API group and their type will be injected into that group by the kube-apiserver.
+
+The *.k8s.io or *.kubernetes.io groups are owned by the Kubernetes community and protected by API review (see [What APIs need to be reviewed](https://github.com/kubernetes/community/blob/master/sig-architecture/api-review-process.md#what-apis-need-to-be-reviewed),
+to ensure consistency and quality.  To avoid confusion in our API groups and prevent accidentally claiming a
+space inside of the kubernetes API groups, the kube-apiserver needs to be updated to protect these reserved API groups.
+
+This KEP proposes adding an `api-approved.kubernetes.io` annotation to CustomResourceDefinition.  This is only needed if
+the CRD group is `k8s.io`, `kubernetes.io`, or ends with `.k8s.io`, `.kubernetes.io`.  The value should be a link to a
+to a URL where the current spec was approved, so updates to the spec should also update the URL.
+ 
+```yaml
+metadata:
+  annotations:
+    "api-approved.kubernetes.io": "https://github.com/kubernetes/kubernetes/pull/78458"
```

Comment:
Why would you hard-code the list vs. an annotation?

Reply:
not sure I understand the question... are you asking if we could hard-code the list of allowed k8s.io CRDs in the kube-apiserver? baking in knowledge of specific APIs into the CRD server is a non-goal

Reply:
> The value should be a link to a to a URL where the current spec was approved, so updates to the spec should also update the URL.

would the server fetch the contents of `https://github.com/kubernetes/kubernetes/pull/78458` to determine if the API was approved?

Reply:
nope, this is primarily intended to 1) prevent accidental use of k8s.io-suffixed APIs and 2) give consumers of CRD manifests a way to check if what they're using matched what was approved (a way to go to the authoritative source)

Reply:
i guess one can cheat by providing a fake PR link, but it's up to the operator catch that.